src/main/java/com/turn/ttorrent/common/Torrent.java

```diff
@@ -418,6 +418,12 @@ public void save(OutputStream output) throws IOException {
 		return md.digest();
 	}
 
+	public static byte[] hash(ByteBuffer data) throws NoSuchAlgorithmException {
+		MessageDigest md = MessageDigest.getInstance("SHA-1");
+		md.update(data);
```

Comment:
Hey, I think I can sum the bytes duplication:

- each byte piece is duplicated from the piecesHashes -> this one you got rid of by hitting the piecesHashes using the offset;

- each file piece is read into a ByteBuffer then copied into a byte array -> this one you got rid of copying, by passing the ByteBuffer (which has a byte array internally) to the MessageDigest mehod;

I *think* that what remains is limiting the file reading into ByteBuffer, like I tried in #195 , I'll try to limit it then profile it, with this done I think this patch is complete.